### PR TITLE
Tweak overheads of Regex cache access

### LINKED ISSUE #50051: Regressions in System.Text.RegularExpressions.Tests.Perf_Regex_Cache
### Run Information

Architecture | x86
-- | --
OS | Windows 10.0.18362
Baseline | [5d984be87c70b71136efa8334a503c56cbcf2ccb](https://github.com/dotnet/runtime/commit/5d984be87c70b71136efa8334a503c56cbcf2ccb)
Compare | [e0bbfa0f555cf5b9d8b8749e1a91a78924c9d927](https://github.com/dotnet/runtime/commit/e0bbfa0f555cf5b9d8b8749e1a91a78924c9d927)

### Regressions in System.Text.RegularExpressions.Tests.Perf_Regex_Cache

Benchmark | Baseline | Test | Test/Base | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | --
[IsMatch](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x86_Windows 10.0.18362/System.Text.RegularExpressions.Tests.Perf_Regex_Cache.IsMatch(total%3a%2040000%2c%20unique%3a%201600%2c%20cacheSize%3a%203200).html>) | 19.42 ms | 20.49 ms | 1.06 | | | | [Trace](https://helixri107v0xdeko0k025g8.blob.core.windows.net/results-c55f40b6eb0d465b90/Collect%20System.Text.RegularExpressions.Tests.Perf_Regex_Cache.IsMatch/artifacts/BenchmarkDotNet.Artifacts/System.Text.RegularExpressions.Tests.Perf_Regex_Cache.IsMatch%28total_%2040000%2C%20unique_%201600%2C%20cacheSize_%203200%29-20210310-121737.etl?sv=2019-07-07&se=2021-06-08T20%3A15%3A48Z&sr=c&sp=rl&sig=z%2FVFAC%2FSrMFlb90k9tOsmcZajLqVZYITuZKlYce5enU%3D) | [Trace](https://helixri107v0xdeko0k025g8.blob.core.windows.net/results-3a33f843f2164fa98b/Collect%20System.Text.RegularExpressions.Tests.Perf_Regex_Cache.IsMatch/artifacts/BenchmarkDotNet.Artifacts/System.Text.RegularExpressions.Tests.Perf_Regex_Cache.IsMatch%28total_%2040000%2C%20unique_%201600%2C%20cacheSize_%203200%29-20210310-121854.etl?sv=2019-07-07&se=2021-06-08T20%3A15%3A48Z&sr=c&sp=rl&sig=%2Fx4NWzMPPStYAqIGg5Esgi%2BUwc0Au4bxkO7ESLYOsnw%3D)

graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/03_10_2021/refs/heads/main_x86_Windows%2010.0.18362_Regression/System.Text.RegularExpressions.Tests.Perf_Regex_Cache.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
py .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'System.Text.RegularExpressions.Tests.Perf_Regex_Cache*'
```
<details>

.
### Payloads

[Baseline](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-0e1af89b-8d38-4344-9eb2-58e62fbb02f12e0dc689ebc4a20aa/00f25d3f-8830-4d35-9170-b006a086eb0c.zip?sv=2019-07-07&se=2021-04-08T12%3A37%3A18Z&sr=c&sp=rl&sig=AoqklEw1Mi7SPBssm8yjsvokT9EFH0NKUxD6dPol%2FCk%3D>)
[Compare](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-bb230eb9-e046-4123-8388-17ef08ba304968b81cebc3446c8b6/65499052-1456-4dc3-a705-a308f1b0cfe8.zip?sv=2019-07-07&se=2021-04-08T19%3A51%3A19Z&sr=c&sp=rl&sig=C08I039d17m4b31uxV5Z9vH1ZF1p6e%2B%2FTT8Ui8%2Fk3%2BQ%3D>)

#### System.Text.RegularExpressions.Tests.Perf_Regex_Cache.IsMatch(total: 40000, unique: 1600, cacheSize: 3200)

```log
[18440759.175 ; 18999711.597) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[18999711.597 ; 19660073.509) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[19660073.509 ; 20370924.023) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[20370924.023 ; 21163475.753) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[21163475.753 ; 21619265.718) | 
[21619265.718 ; 22177071.195) | @
```
### Baseline Jit Disasm

```assembly
; BenchmarkDotNet.Autogenerated.Runnable_3.__ForDisassemblyDiagnoser__()
       push      ebp
       mov       ebp,esp
       sub       esp,10
       xor       eax,eax
       mov       [ebp+0FFF8],eax
       mov       [ebp+0FFF4],eax
       mov       [ebp+0FFF0],eax
       mov       [ebp+0FFFC],ecx
       mov       ecx,[ebp+0FFFC]
       cmp       dword ptr [ecx+34],0B
       jne       short M00_L00
       mov       ecx,[ebp+0FFFC]
       mov       ecx,[ecx+28]
       mov       [ebp+0FFF8],ecx
       mov       ecx,[ebp+0FFFC]
       mov       ecx,[ecx+2C]
       mov       [ebp+0FFF4],ecx
       mov       ecx,[ebp+0FFFC]
       mov       ecx,[ecx+30]
       mov       [ebp+0FFF0],ecx
       mov       ecx,[ebp+0FFF4]
       push      ecx
       mov       ecx,[ebp+0FFF0]
       push      ecx
       mov       ecx,[ebp+0FFFC]
       mov       edx,[ebp+0FFF8]
       call      09D9F148
       mov       esp,ebp
       pop       ebp
       ret
M00_L00:
       xor       eax,eax
       mov       esp,ebp
       pop       ebp
       ret
; Total bytes of code 85
```

### Compare Jit Disasm

```assembly
; System.Text.RegularExpressions.Tests.Perf_Regex_Cache.IsMatch(Int32, Int32, Int32)
       push      ebp
       mov       ebp,esp
       push      edi
       push      esi
       push      ebx
       mov       edi,ecx
       mov       esi,edx
       mov       ecx,[ebp+8]
       cmp       ds:[89E0],ecx
       je        short M00_L00
       test      ecx,ecx
       jl        short M00_L01
       call      System.Text.RegularExpressions.RegexCache.set_MaxCacheSize(Int32)
M00_L00:
       mov       ecx,[edi+4]
       mov       ebx,[ebp+0C]
       push      ebx
       push      esi
       call      dword ptr ds:[2F4]
       push      esi
       push      eax
       mov       ecx,edi
       xor       edx,edx
       call      System.Text.RegularExpressions.Tests.Perf_Regex_Cache.RunTest(Int32, Int32, System.String[])
       pop       ebx
       pop       esi
       pop       edi
       pop       ebp
       ret       8
M00_L01:
       mov       ecx,11
       call      System.Text.RegularExpressions.ThrowHelper.ThrowArgumentOutOfRangeException(System.Text.RegularExpressions.ExceptionArgument)
       int       3
; Total bytes of code 73
```
```assembly
; System.Text.RegularExpressions.RegexCache.set_MaxCacheSize(Int32)
       push      ebp
       mov       ebp,esp
       push      edi
       push      esi
       push      ebx
       sub       esp,40
       vzeroupper
       vxorps    xmm4,xmm4,xmm4
       vmovdqu   xmmword ptr [ebp+0FFC0],xmm4
       xor       eax,eax
       mov       [ebp+0FFD0],eax
       mov       [ebp+0FFD4],eax
       mov       [ebp+0FFE8],eax
       mov       ebx,ecx
       mov       edx,ds:[57D0]
       mov       eax,edx
       mov       [ebp+0FFBC],eax
       xor       edx,edx
       mov       [ebp+0FFDC],edx
       cmp       byte ptr [ebp+0FFDC],0
       jne       near ptr M01_L07
       lea       edx,[ebp+0FFDC]
       mov       ecx,eax
       call      System.Threading.Monitor.ReliableEnter(System.Object, Boolean ByRef)
       mov       ds:[89E0],ebx
       test      ebx,ebx
       jne       short M01_L01
       mov       ecx,ds:[57D0]
       cmp       [ecx],ecx
       call      dword ptr ds:[315C]
       mov       ecx,ds:[57D4]
       inc       dword ptr [ecx+0C]
       mov       edx,[ecx+8]
       xor       eax,eax
       mov       [ecx+8],eax
       test      edx,edx
       jle       short M01_L00
       mov       ecx,[ecx+4]
       push      edx
       xor       edx,edx
       call      System.Array.Clear(System.Array, Int32, Int32)
M01_L00:
       mov       dword ptr ds:[57CC],0
       mov       dword ptr [ebp+0FFE4],0
       mov       dword ptr [ebp+0FFE8],0FC
       push      offset M01_L10
       jmp       near ptr M01_L11
M01_L01:
       mov       ecx,ds:[57D4]
       cmp       ebx,[ecx+8]
       jge       near ptr M01_L09
       mov       esi,ds:[57D4]
       mov       ecx,0A6C8968
       mov       edx,22
       call      CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
       mov       edx,ds:[57E0]
       test      edx,edx
       jne       short M01_L02
       mov       ecx,offset MT_System.Comparison`1[[System.Text.RegularExpressions.RegexCache+Node, System.Text.RegularExpressions]]
       call      CORINFO_HELP_NEWSFAST
       mov       ecx,ds:[57DC]
       test      ecx,ecx
       je        near ptr M01_L08
       lea       edx,[eax+4]
       call      CORINFO_HELP_ASSIGN_REF_ECX
       mov       dword ptr [eax+0C],0AB7AF78
       lea       edx,ds:[57E0]
       call      CORINFO_HELP_CHECKED_ASSIGN_REF_EAX
       mov       edx,eax
M01_L02:
       mov       ecx,esi
       cmp       [ecx],ecx
       call      System.Collections.Generic.List`1[[System.__Canon, System.Private.CoreLib]].Sort(System.Comparison`1<System.__Canon>)
       mov       edx,ds:[57D4]
       mov       [ebp+0FFB4],edx
       cmp       dword ptr [edx+8],0
       jbe       near ptr M01_L06
       lea       edx,ds:[57CC]
       mov       ecx,[ebp+0FFB4]
       mov       ecx,[ecx+4]
       cmp       dword ptr [ecx+4],0
       jbe       short M01_L04
       mov       ecx,[ecx+8]
       call      CORINFO_HELP_CHECKED_ASSIGN_REF_ECX
       mov       esi,ebx
M01_L03:
       mov       ecx,ds:[57D4]
       mov       edx,ecx
       mov       eax,[edx+8]
       cmp       esi,eax
       jge       short M01_L05
       mov       edx,ds:[57D0]
       cmp       esi,eax
       jae       short M01_L06
       mov       ecx,[ecx+4]
       cmp       esi,[ecx+4]
       jae       short M01_L04
       mov       [ebp+0FFD8],esi
       mov       eax,[ecx+esi*4+8]
       lea       esi,[eax+10]
       lea       edi,[ebp+0FFC0]
       mov       ecx,5
       rep movsd
       mov       [ebp+0FFB8],edx
       cmp       [edx],edx
       push      [ebp+0FFD0]
       push      [ebp+0FFCC]
       push      [ebp+0FFC8]
       push      [ebp+0FFC4]
       push      [ebp+0FFC0]
       push      0
       push      0
       lea       edx,[ebp+0FFD4]
       mov       ecx,[ebp+0FFB8]
       call      System.Collections.Concurrent.ConcurrentDictionary`2[[System.Text.RegularExpressions.RegexCache+Key, System.Text.RegularExpressions],[System.__Canon, System.Private.CoreLib]].TryRemoveInternal(Key, System.__Canon ByRef, Boolean, System.__Canon)
       mov       esi,[ebp+0FFD8]
       inc       esi
       jmp       short M01_L03
M01_L04:
       call      CORINFO_HELP_RNGCHKFAIL
M01_L05:
       sub       eax,ebx
       push      eax
       mov       edx,ebx
       cmp       [ecx],ecx
       call      System.Collections.Generic.List`1[[System.__Canon, System.Private.CoreLib]].RemoveRange(Int32, Int32)
       jmp       short M01_L09
M01_L06:
       call      System.ThrowHelper.ThrowArgumentOutOfRange_IndexException()
       int       3
M01_L07:
       call      System.Threading.Monitor.ThrowLockTakenException()
       int       3
M01_L08:
       call      System.MulticastDelegate.ThrowNullThisInDelegateToInstance()
       int       3
M01_L09:
       cmp       byte ptr [ebp+0FFDC],0
       je        short M01_L13
       mov       ecx,[ebp+0FFBC]
       call      System.Threading.Monitor.Exit(System.Object)
       jmp       short M01_L13
M01_L10:
       mov       dword ptr [ebp+0FFE8],0
       jmp       short M01_L13
M01_L11:
       cmp       byte ptr [ebp+0FFDC],0
       je        short M01_L12
       mov       ecx,[ebp+0FFBC]
       call      System.Threading.Monitor.Exit(System.Object)
M01_L12:
       pop       eax
       jmp       eax
M01_L13:
       lea       esp,[ebp+0FFF4]
       pop       ebx
       pop       esi
       pop       edi
       pop       ebp
       ret
; Total bytes of code 495
```
```assembly
; System.Text.RegularExpressions.Tests.Perf_Regex_Cache.RunTest(Int32, Int32, System.String[])
       push      ebp
       mov       ebp,esp
       push      edi
       push      esi
       push      ebx
       sub       esp,8
       mov       esi,edx
       mov       ebx,[ebp+0C]
       mov       edi,[ebp+8]
       xor       eax,eax
       mov       [ebp+0FFF0],eax
       xor       edx,edx
       test      ebx,ebx
       jle       short M02_L03
M02_L00:
       mov       [ebp+0FFEC],edx
       lea       ecx,[esi+edx]
       cmp       ecx,[edi+4]
       jae       short M02_L04
       mov       ecx,[edi+ecx*4+8]
       call      System.Text.RegularExpressions.RegexCache.GetOrAdd(System.String)
       mov       ecx,eax
       mov       edx,ds:[3B88]
       cmp       [ecx],ecx
       call      System.Text.RegularExpressions.Regex.IsMatch(System.String)
       movzx     eax,al
       xor       eax,[ebp+0FFF0]
       movzx     eax,al
       mov       edx,[ebp+0FFEC]
       inc       edx
       cmp       edx,ebx
       jl        short M02_L02
M02_L01:
       lea       esp,[ebp+0FFF4]
       pop       ebx
       pop       esi
       pop       edi
       pop       ebp
       ret       8
M02_L02:
       mov       [ebp+0FFF0],eax
       jmp       short M02_L00
M02_L03:
       mov       eax,[ebp+0FFF0]
       jmp       short M02_L01
M02_L04:
       call      CORINFO_HELP_RNGCHKFAIL
       int       3
; Total bytes of code 106
```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### PR DESCRIPTION
**Pull request by stephentoub**
Small improvement, but then the cited regression was also small.  I tried a few other things (e.g. passing the Key by in, storing an int milliseconds timeout instead of the TimeSpan, etc.), but they didn't help, and I'm not seeing a ton more that can be done. Closes https://github.com/dotnet/runtime/issues/50051.

This is on 64-bit:

|  Method |         Toolchain |  total | unique | cacheSize |     Mean |    Error |   StdDev |   Median |      Min |      Max | Ratio |
|-------- |------------------ |------- |------- |---------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|
| IsMatch | \main\corerun.exe |  40000 |   1600 |      3200 | 11.26 ms | 0.099 ms | 0.083 ms | 11.24 ms | 11.08 ms | 11.40 ms |  1.00 |
| IsMatch |   \pr\corerun.exe |  40000 |   1600 |      3200 | 11.08 ms | 0.177 ms | 0.165 ms | 11.06 ms | 10.87 ms | 11.40 ms |  0.99 |
|         |                   |        |        |           |          |          |          |          |          |          |       |
| IsMatch | \main\corerun.exe | 400000 |      1 |        15 | 43.40 ms | 0.285 ms | 0.238 ms | 43.37 ms | 43.00 ms | 43.85 ms |  1.00 |
| IsMatch |   \pr\corerun.exe | 400000 |      1 |        15 | 41.89 ms | 0.215 ms | 0.180 ms | 41.84 ms | 41.66 ms | 42.37 ms |  0.97 |
|         |                   |        |        |           |          |          |          |          |          |          |       |
| IsMatch | \main\corerun.exe | 400000 |      7 |        15 | 57.78 ms | 0.367 ms | 0.343 ms | 57.65 ms | 57.30 ms | 58.45 ms |  1.00 |
| IsMatch |   \pr\corerun.exe | 400000 |      7 |        15 | 54.56 ms | 0.277 ms | 0.259 ms | 54.54 ms | 54.13 ms | 55.02 ms |  0.94 |

### CODE REVIEW

**Review by phocodes — APPROVED** [review]

**Comment by stephentoub** [comment]
@CertifiedRice, you "approved" about 25 PRs in rapid succession.  Did you actually review them?  Can you help me understand what action you took that led you to "approve" them?  Thanks.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Cache.cs [hunk 1]
```diff
@@ -139,20 +141,18 @@ public static Regex GetOrAdd(string pattern, RegexOptions options, TimeSpan matc
...
 
             // We optimize for repeated usage of the same regular expression over and over,
             // by having a fast-path that stores the most recently used instance.  Check
             // to see if that instance is the one we want; if it is, we're done.
-            Node? lastAccessed = s_lastAccessed;
-            if (lastAccessed != null)
+            if (s_lastAccessed is Node lastAccessed)
             {
-                if (lastAccessed.Key.Equals(key))
+                if (key.Equals(lastAccessed.Key))
```
Comment (eerhardt) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Cache.cs:153]: For my knowledge - did this line make a perf difference? Or does it just read better this way?

**Review by eerhardt — APPROVED** [review]
Looks good to me.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Cache.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Cache.cs:153]: It did for an intermediate stage where I was experimenting with passing key as `in`.  I ended up reverting that, but left this as it was as I thought it still read slightly better.

**Review by pgovind — APPROVED** [review]
The code change LGTM. But, I don't really get why this is performing slightly better :/

**Comment by eerhardt** [comment]
> But, I don't really get why this is performing slightly better :/

I'll take a stab at answering, and @stephentoub can correct me 😄. It basically comes to 2 micro-optimizations:

1. We are no longer considering `_culture.GetHashCode()` in the hash code of the `Key`.
2. Even a smaller optimization is that we don't need to pass back 2 values from `TryGet` - a `bool` and a `Regex`. Instead, we are just passing back the `Regex`. Honestly, I'm not sure myself exactly _why_ that part makes a measurable change....

**Comment by pgovind** [comment]
> We are no longer considering _culture.GetHashCode() in the hash code of the Key.

Yup, I noticed that part. We're reducing the amount of computation here, so this makes sense.

> that we don't need to pass back 2 values from TryGet - a bool and a Regex.

This is the part that I'm surprised by. The returned bool was on the stack, so I can't think of any reason why this would make a measurable change. Maybe when there's some down time I'll look at the IL generated here and/or measure how much this change contributed to the overall speedup. I'm guessing all of the speedup here is from point 1?

**Comment by eerhardt** [comment]
Doing a really small example:

https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgAwAEmuAdACoCmAHglQEp0DmArgDYQBOAUQYAHPnRQooAexgoA3ESKYAzBWxkAwmQDeRMvrKioANwgI6FXADYybdozIoA+rxQIAggGNP4lHQAmCoQGZHoGKrYcDgDidAgAFACUOmEh+lAAZmTxzq4e3r4BZFAokfYMZGLlialpusFpjZgA7JVRDEGN+gC+tSEtZDA83J0GvYS1EcBSUtxkNHwAnrEJUpwIZQ58yfVdxVk5Lqj5PhJFJZsVVYw1DY27ewZ8ZAC8beWjjxStCHycdJ80uM9s83kNuCM+uFWhkINw/IDxt0gA

```
C.Get()
    L0000: mov ecx, 0x17e5c5f0
    L0005: xor edx, edx
    L0007: call 0x5c2ffde0
    L000c: mov eax, [eax]
    L000e: test eax, eax
    L0010: je short L0013
    L0012: ret
    L0013: xor eax, eax
    L0015: ret

C.TryGet(System.Text.RegularExpressions.Regex ByRef)
    L0000: push esi
    L0001: mov esi, edx
    L0003: mov ecx, 0x17e5c5f0
    L0008: xor edx, edx
    L000a: call 0x5c2ffde0
    L000f: mov eax, [eax]
    L0011: test eax, eax
    L0013: je short L0023
    L0015: mov edx, esi
    L0017: call 0x5c1d7098
    L001c: mov eax, 1
    L0021: pop esi
    L0022: ret
    L0023: xor eax, eax
    L0025: mov [esi], eax
    L0027: pop esi
    L0028: ret
```

**Comment by stephentoub** [comment]
There's a difference between handing back an object via a return and handing back an object via a ref... the latter requires a write barrier:
```assembly
; Program.ByReturn()
       mov       rax,[rcx+8]
       ret
; Total bytes of code 5

; Program.ByOut(System.Object ByRef)
       mov       rax,rdx
       mov       rdx,[rcx+8]
       mov       rcx,rax
       call      CORINFO_HELP_CHECKED_ASSIGN_REF
       nop
       ret
; Total bytes of code 17
```

Our Try pattern goals afoul of this for usability reasons, but you can see for example as a minor optimization our internal abstraction in LINQ switches the bool/T positions so that when T is a reference type it's returned rather than passed out by ref:
https://github.com/dotnet/runtime/blob/7e651850f7eeb69fa5e9a291197f22a9721a48a1/src/libraries/System.Linq/src/System/Linq/IPartition.cs#L33-L47

**Pull request merged**